libbeat/cmd/instance/beat.go

```diff
@@ -696,8 +707,69 @@ func (b *Beat) registerTemplateLoading() error {
 			return err
 		}
 
-		if esCfg.Index != "" && (cfg.Name == "" || cfg.Pattern == "") && (b.Config.Template == nil || b.Config.Template.Enabled()) {
-			return fmt.Errorf("setup.template.name and setup.template.pattern have to be set if index name is modified.")
+		if esCfg.Index != "" && (templateCfg.Name == "" || templateCfg.Pattern == "") && (b.Config.Template == nil || b.Config.Template.Enabled()) {
+			return fmt.Errorf("setup.template.name and setup.template.pattern have to be set if index name is modified")
+		}
+
+		if b.Config.ILM.Enabled() {
+
+			cfgwarn.Beta("Index lifecycle management is enabled which is in beta.")
+			logp.Info("Overwriting setup.template for ILM")
+
+			ilmCfg, err := getILMConfig(b)
+			if err != nil {
+				return err
+			}
+
+			// Template name and pattern can't be configure when using ILM
+			logp.Info("Overwriting template name and pattern as ILM is enabled.")
+			b.Config.Template.SetString("name", -1, ilmCfg.RolloverAlias)
+			b.Config.Template.SetString("pattern", -1, fmt.Sprintf("%s-*", ilmCfg.RolloverAlias))
+
+			// rollover_alias and lifecycle.name can't be configured and will be overwritten
+			logp.Info("Overwrite template settings for rollover_alias and lifecycle.name.")
+			err = b.Config.Template.SetString("settings.index.lifecycle.rollover_alias", -1, ilmCfg.RolloverAlias)
+			if err != nil {
+				logp.Err("error setting rollover_alias setting: %s", err)
+			}
+			err = b.Config.Template.SetString("settings.index.lifecycle.name", -1, ILMPolicyName)
+			if err != nil {
+				logp.Err("error setting index lifecycle name setting err: %s", err)
+			}
```

Comment:
Why are there two different configuration settings (`output.elasticsearch.ilm*` and `settings.index.lifecycle`)  when `settings.index.lifecycle.*` is always overwritten? Also where is it used or can it be configured, I can't find it anywhere else.

Reply:
The values you see above are put into the index template. These "could" be set through `setup.template.settings.*` assuming someone would not use the "out of the box" ILM version. Having the policy name in the template will make sure ILM will know which policy must be applied to the index.

Perhaps I miss something above as I didn't fully understand the question.